Add simplify function for predict() and evaluate()

terrytangyuan opened the pull request:
This addresses https://github.com/rstudio/tfestimators/issues/49. @jjallaire This is a little bit different than the original approach you outlined:

*  "We provide a hook for custom estimators to provide a simplifier (perhaps this is attached to model_fn, perhaps via an S3 method)" - instead of doing this, here we are taking directly from the predictions result and apply a function to them. I think this is the easier to maintain and easiest for users to follow to write their own functions. 

* Users should be able to provide their own function for simplification. For package maintainer or canned estimator implementors, they can write a method like `predict.tf_estimator_dnn_linear_combined_classifier()` to override the existing `predict.tf_estimator()` since a canned estimator has multiple classes (I need to experiment to see what the order/priority of execution would be for an object with multiple classes in order to do this correctly):
```
> class(clf)
[1] "tf_estimator"                                "tf_estimator_classifier"                    
[3] "tf_estimator_dnn_linear_combined_classifier"
```
This allows the implementation of `predict()` to be model type level (multihead, time series, clustering) instead of per canned estimator level. I am uncertain whether this is the right approach so I opened this PR for comments and suggestions. 

* Should the default of `simplify` be `FALSE` for backward compatibility purposes? From another perspective, set default to be `TRUE` lowers the entry of barrier for a lot of users, especially new users.

jjallaire commented:
We don't have enough users to worry about the backward compatibility issue so I think the default should definitely be `simplify = TRUE`

jjallaire commented on DESCRIPTION:
```diff
@@ -34,7 +34,9 @@ Imports:
-    utils
+    utils,
+    purrr,
+    dplyr
```
I think we can get what you are looking for from dplyr by just importing the `tibble` package (it has row binding and of course as_tibble)

kevinushey replied:
I thought `bind_rows()` was `dplyr`-specific?

jjallaire replied:
There is definitely row binding code in tibble (I use it in tfruns). You can get as_tibble and glimpse from tibble (http://tibble.tidyverse.org/reference/index.html). Not sure about bind_rows but I wonder if add_row would suffice?

terrytangyuan replied:
Yes, those two functions can be imported from tibble. I tried to use `add_row` but it doesn't seem as efficient as `bind_rows` since the latter is implemented in C++.

jjallaire replied:
Okay, sounds good.

Also, we may need to update some of our documentation to reflect `simplify = TRUE` as part of the PR.

Let us know once you think this PR is ready for final review and merge.

jjallaire commented:
Will do a detailed review later today.

jjallaire commented on R/tf_estimator.R:
```diff
@@ -16,6 +16,10 @@ tf_classifier <- function(estimator, class, args = NULL) {
 }
 
+is.tf_custom_estimator <- function(object) {
+  "tf_custom_estimator" %in% class(object)
```
You can do `inherits(object, "tf_custom_estimator")` for this

jjallaire commented on R/tf_estimator.R:
```diff
@@ -141,6 +154,7 @@ predict.tf_estimator <- function(object,
                                  predict_keys = NULL,
                                  hooks = NULL,
                                  as_iterable = FALSE,
+                                 simplify = TRUE,
```
I think that TRUE should really mean "auto", which is to use a simplifier for canned estimators and to use a simplifier for custom estimators *when it is available". There should be a way for custom estimators to include a simplifier in their definition (either something returned from model_fn or an S3 method, whichever makes the most sense).

terrytangyuan replied:
Could you elaborate a little more why it's necessary to include this in their definition? This simplifier is currently only used in prediction phase. Do you mean we should allow users to specify this in two possible places? It might be better to make it consistent for both canned and custom estimators to specify simplifier in one place.

kevinushey replied:
I think there's effectively three behaviors users might want at this level:

`TRUE`: Use the simplifer function associated with an estimator (if any).
`FALSE`: Don't simplify anything; return the 'base' output as from TensorFlow.
A function: Use this function in lieu of any registered simplifier on the estimator. (This use case is somewhat more esoteric)

jjallaire replied:
I guess I was just thinking that a "custom" estimator might actually be a library that someone has created for others to use. In that case I don't think the end users should be burdened with having to learn a special syntax (pass a specific simplifier function) to use that estimator the same way they use canned estimators (which get simplified output automatically).

I guess I'm just asking that custom estimators be able to appear to the end user like canned estimators. I realize that if the developer of a custom estimator is the same one calling predict it's just as straightforward to pass a simplify function, but for their end users I think it's better if it's somehow bound to the definition of the custom estimator.

terrytangyuan replied:
I think in most cases the developer of a custom estimator should be the same one calling predict. Otherwise, it would be a canned estimator. Assume there's a `xxx_estimator`, developer should then implement a `predict.xxx_estimator()` with a custom simplifier targeting this new estimator. Then users would only need to call `predict(xxx_estimator_object, simplify = TRUE)`. 

This simplifier is too specific to the prediction phase that it's better not to add it as part of the estimator definition. Estimator definition should only contain the model definitions and parameters. We can think of this simplifier as `"output_fn"` that should be tied to the specific actions on the model (e.g. train, predict, evaluate). 

There could potentially be a `evaluation_simplifier` in the future that we only want to add as part of `evaluate()`. Otherwise in the model definition there will be multiple simplifiers. We can see these as part of different specs (e.g. predict_spec, eval_spec) similar to what's in my `train_and_evaluate()` PR. This will prevent the parameters in the model definition from exploding.

jjallaire replied:
Okay, I buy all that. Thanks for bearing with me on this one, I'm now convinced we've got the right approach.

jjallaire commented on R/simplify_fn.R:
```diff
@@ -0,0 +1,36 @@
+  if (is.function(simplify) && validate_prediction_simplify_fn(simplify)) {
+    simplify(predictions)
+  } else {
+    if (isTRUE(simplify)) {
```
Again, `TRUE` should mean "auto", which means we'll just use whatever simplifier is registered for custom estimators or bypass it entirely if there is none. 

I think we should keep the argument as `TRUE` rather than changing it to "auto" as I think the logical is simpler and more readily understood. `simplify = TRUE` really means "simplify if possible" .

jjallaire commented:
Let's add a "Generating Predictions" Vignette to the this PR which:

1) Shows some sample output from various canned estimators

2) Introduces the concept of `simplify = TRUE` and shows some sample "unsimplified" output.

3) Illustrates how to use a custom simplifier by passing a function.

4) Shows how to register a simplifier for a custom estimator.

Once we have this we can solicit more feedback from Hadley, Max, Nathan, Edgar, Sean, etc.

kevinushey commented on R/simplify_fn.R:
```diff
@@ -0,0 +1,36 @@
+}
+
+validate_prediction_simplify_fn <- function(prediction_simplify_fn) {
+  if (length(formals(prediction_simplify_fn)) == 1) {
```
I think this would fail for simplify functions that take optional arguments, which might be unintended. E.g.

```
simplify <- function(x, how = "method") { ... }
```

kevinushey replied:
You might want to just validate that it's a function, and let it fail at runtime if it fails to invoke for whatever reason.

terrytangyuan replied:
Sounds good I'll remove that check

terrytangyuan commented:
@jjallaire I think it's ready for another round of review. I only pushed the raw Rmd file for now since I want to double check with you to see whether it's appropriate to put this in "advanced" tutorials. Also I imported `as_tibble` and `glimpse` from `dplyr` on purpose so the description will look shorter (`dplyr` is being used anyways for `bind_rows`).

jjallaire commented on vignettes/simplify_function.Rmd:
```diff
@@ -1,28 +1,121 @@
+## Overview
 
-This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.
+Estimator's `predict()` and `evaluate()` provide an easy and efficient way of making predictions and evaluations on new data. However, the prediction and evaluation results are pretty verbose and usually in forms of nested lists, which isn't user-friendly. By default, the results produced from canned estimators will be simplified using a default simplify function for both `predict()` and `evaluat ...
```
"can also write a custom simplify function for them"

jjallaire replied:
"illustrate what the simplified results look like"

jjallaire commented on vignettes/simplify_function.Rmd:
````diff
@@ -1,28 +1,121 @@
+model %>% train(mtcars_input_fn(train_data))
+```
+
+Here's how the prediction output looks like without applying the simplifier on the results:
````
"Here's what"

jjallaire commented on vignettes/simplify_function.Rmd:
````diff
@@ -1,28 +1,121 @@
+model %>% predict(mtcars_input_fn(mtcars[1, ]), simplify = FALSE)
+```
+
+Note that here we only evaluate the first row of `mtcars` data since the results will be in forms of long nested lists that may not fit in this page.
````
We can remove this qualifier I think

jjallaire commented on vignettes/simplify_function.Rmd:
```diff
@@ -1,28 +1,121 @@
+
+### Default Simplifier
+
+If `simplify = TRUE`, a default simplify function for predictions will be used to flatten the results so they are more appearing and concise like the following:
```
"more appealing"

jjallaire replied:
We should paste the output directly into the vignette so that the vignettes can be built on a system without TensorFlow (and also so that they build faster)

jjallaire commented on vignettes/simplify_function.Rmd:
```diff
@@ -1,28 +1,121 @@
+
+### Custom Simplifier for Prediction Result
+
+You can also write your own custom simplify function to clean up the prediction results. For example, below we define a custom function that obtains the probabilities from the list of prediction results with keys being the prediction keys pre-defined for this canned estimator. In this case, `linear_classifier` has logits, probabilities, prediction classes, logistics, etc. pre-defined. We can see  ...
```
"clean up prediction results"

jjallaire replied:
I thought we were going to move away from the enum style "prediction_keys" and instead just use simple character vectors as is customary elsewhere in R?

jjallaire commented on vignettes/simplify_function.Rmd:
```diff
@@ -1,28 +1,121 @@
+
+## Simplify Evaluation Results
+
+Similarly, the evaluation results can be simplified as well. Here's how the raw evaluation results look like without simplification by providing `simplify = FALSE`. 
```
"Evaluation results can also be simplified. Here's what the raw evaluation results look like without simplification.."

jjallaire commented on vignettes/simplify_function.Rmd:
```diff
@@ -1,28 +1,121 @@
+
+## Writing Simplify Function for Custom Estimators
+
+So far we've only illustrated how to simplify the results for canned estimator, in particular, `linear_classifier`. Custom estimators do not simplify results by default unless you provide a custom simplify function. The way to write them is almost identical to canned estimators. The predictions and evaluations results are highly dependent on what's defined in the `model_fn` so you need to pay att ...
```
"for canned estimators"

jjallaire commented on vignettes/simplify_function.Rmd:
```diff
@@ -1,28 +1,121 @@
+predictions <- list(
+    class = tf$argmax(logits, 1L),
+    prob = tf$nn$softmax(logits))
+if (mode == mode_keys()$PREDICT) {
```
I think we should be using straight character literals here rather than the enum style.

terrytangyuan replied:
How about using `mode_keys()$PREDICT` for development but using "infer" - (straight character literal) - in tutorials? This way it's easier for users and easier for maintenance.

jjallaire commented:
My only question about the use of `bind_rows()` is whether we really need it for performance (as it implies that we take on dplyr as a dependency, which is as dependencies go on the heavier side).

jjallaire commented:
I don't mind using those in development. Note when providing argument
defaults for enums we should use this form:

predict_keys = c("classes", "probs", ...)

On Thu, Oct 26, 2017 at 11:43 AM, Yuan (Terry) Tang <
notifications@github.com> wrote:

> *@terrytangyuan* commented on this pull request.
> ------------------------------
>
> In vignettes/simplify_function.Rmd
> <https://github.com/rstudio/tfestimators/pull/106#discussion_r147183847>:
>
> > +model %>% evaluate(mtcars_input_fn(test_data), simplify = simplify_fn)
> +```
> +
> +
> +## Writing Simplify Function for Custom Estimators
> +
> +So far we've only illustrated how to simplify the results for canned estimator, in particular, `linear_classifier`. Custom estimators do not simplify results by default unless you provide a custom simplify function. The way to write them is almost identical to canned estimators. The predictions and evaluations results are highly dependent on what's defined in the `model_fn` so you need to pay attention to them when writing the simplify function.
> +
> +For example, below is a snippet in the `model_fn` that defines the behavior for prediction mode (note the `mode == mode_keys()$PREDICT`). `predictions` is being returned during this phase and is a list of keyed items for raw prediction classes and probabilities.
> +
> +
> +```{r eval=FALSE}
> +predictions <- list(
> +    class = tf$argmax(logits, 1L),
> +    prob = tf$nn$softmax(logits))
> +if (mode == mode_keys()$PREDICT) {
>
> How about using mode_keys()$PREDICT for development but using "infer" -
> (straight character literal) - in tutorials? This way it's easier for users
> and easier for maintenance.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rstudio/tfestimators/pull/106#discussion_r147183847>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGXxz8QNEK0VNKzrI-0ZHmC_Wi0-ZChks5swKiUgaJpZM4P5HNH>
> .
>

terrytangyuan commented:
@jjallaire I've addressed your comments in https://github.com/rstudio/tfestimators/pull/106/commits/84d10b2471f5b9dc946d9c65389848f0972d0f95 (the whole diff gets really large after I regenerated the website). I'll play around a bit more regarding `bind_rows`.

terrytangyuan commented:
@jjallaire I've removed `bind_rows()` in https://github.com/rstudio/tfestimators/pull/106/commits/4cc382f58f87c56ac7cd7a1d81d6de6535b0823f. 

cc: @edgararuiz Note that I also removed `map(~.x[[1]])` because we don't want to truncate the probabilities output (in many cases analysts want to use the prediction probabilities for each class).

jjallaire commented:
Terrific!

jjallaire commented:
This PR looks good to merge from my end. Any objection?

kevinushey commented on R/session_run_hooks_custom.R:
```diff
@@ -79,9 +79,9 @@ hook_view_metrics <- function(props, mode_key) {
   
   format <- if (is.null(steps))
-    paste("[:spin]", label, "-- loss: :loss, step: :step")
+    paste("[:spin]", label, "-- loss: :loss, step: :step \n")
```
Why are you adding a newline here? The progress bar needs to redraw the line as new iterations are run; wouldn't this break that (you'd have a new line for each progress redraw)?

terrytangyuan replied:
That is for some testing! Will remove! Thanks for spotting it!

kevinushey commented on R/simplify_fn.R:
```diff
@@ -0,0 +1,37 @@
+  predictions %>%
+    purrr::reduce(rbind) %>%
+    tibble::as_tibble() %>%
+    tibble::glimpse()
```
I think we should leave out the `glimpse()` call here. (This just does some printing to the console, and I think we don't want to implicitly print whenever the user requests predictions -- rather, they should explicitly do that themselves if they want to view the object)

terrytangyuan replied:
Sounds good. I can remove this.

kevinushey commented:
If I understand correctly, custom estimators don't have the notion of a simplify function, then? (What the model functions gives you is what you get and the estimator doesn't provide any built-in means for 'simplifying' whatever form of output you might get).

In other words, we only plan on having simplifiers for canned estimators?

terrytangyuan commented:
@kevinushey Yes for custom estimators, users can provide their own simplify function. 

Now that we are revisiting this and I am more familiar with tibble, etc., I think we can perform this same simplification on custom estimators since `tibble` can store any arbitrary lists that `model_fn` returns.  @kevinushey @jjallaire What do you think?

kevinushey commented:
It might be nice to have `simplify = TRUE` imply that users go from 'list of stuff' to 'tibble of stuff' regardless of the type of estimator used.

jjallaire commented:
I agree that if we could simplify custom estimators to `tibble` that would be a very nice win.

terrytangyuan commented:
Addressed the comments!

jjallaire commented:
You can go ahead and merge when ready!
